common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java

```diff
@@ -0,0 +1,225 @@
+ *   {@link FileModificationPollingMonitor} instead.
+ *
+ * - This class won't work for files that is under a symbolic link (pointing to a directory). This is because the
+ *   Java {@link WatchService} does not raise events for modifications of such files. Consider using
+ *   {@link FileModificationPollingMonitor} instead.
+ *
+ */
+@Slf4j
+public class FileModificationEventWatcher extends Thread implements FileModificationMonitor {
+
+    private static final AtomicInteger THREAD_NUM = new AtomicInteger();
+
+    /**
+     * The path of file to watch.
+     */
+    private final Path pathOfFileToWatch;
+
+    /**
+     * The action to perform when the specified file changes.
+     */
+    private final Consumer<WatchEvent<?>> callback;
+
+    private final UncaughtExceptionHandler uncaughtExceptionalHandler = (t, e) -> logException(e);
+
+    private boolean loopContinuously;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param fileToWatch the file to watch
+     * @param callback    the callback to invoke when a modification to the {@code fileToWatch} is detected
+     * @throws NullPointerException if either {@code fileToWatch} or {@code callback} is null
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     */
+    public FileModificationEventWatcher(@NonNull String fileToWatch, @NonNull Consumer<WatchEvent<?>> callback)
+            throws FileNotFoundException {
+        this(Paths.get(fileToWatch), callback, true);
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param fileToWatch path of the file to watch
+     * @param callback          the callback to invoke when a modification to the {@code fileToWatch} is detected
+     * @param loopContinuously  whether to keep the thread look for file modification after one iteration. This option
+     *                          is useful for testing only.
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     * @throws NullPointerException if either {@code fileToWatch}  or {@code callback} is null
+     */
+    @VisibleForTesting
+    FileModificationEventWatcher(Path fileToWatch, Consumer<WatchEvent<?>> callback,
+                                 boolean loopContinuously)
+            throws FileNotFoundException {
+        super();
+        this.validateInput(fileToWatch, callback);
+
+        this.pathOfFileToWatch = fileToWatch;
+        this.callback = callback;
+        setUncaughtExceptionHandler(uncaughtExceptionalHandler);
+
+        // Set the name for this object/thread for identification purposes.
+        this.setName("file-update-watcher-" + THREAD_NUM.incrementAndGet());
+        this.loopContinuously = loopContinuously;
+    }
+
+    @VisibleForTesting
+    String nameOfFileToWatch() {
```

Comment:
`getWatchedFileName`

Reply:
I would also suggest changing the names of the private fields as well, to make them easier to read.

Reply:
Sure, that makes sense now that you say it. I've modified the names.